File: Components/Strata.Foundation.Core/main/typescript/Utility/Mapped.ts
//////////////////////////////////////////////////////////////////////////////
// Mapped.ts
//////////////////////////////////////////////////////////////////////////////

import {Optional}              from "./Optional";
import {NullPointerException}  from "./NullPointerException";

export
class Mapped<I,O>
{
    private readonly method:    string;
    private readonly input:     I;
    private readonly output:    O | null;
    private readonly throwable: Error | null;

    private
    constructor(method: string,input: I,output: O | null,throwable: Error | null)
    {
        this.method    = method;
        this.input     = input;
        this.output    = output;
        this.throwable = throwable;
    }

    getMethod(): string { return this.method; }

    getInput(): I { return this.input; }

    getOutput(): Optional<O> { return Optional.ofNullable<O>(this.output); }

    getThrowable(): Optional<Error> { return Optional.ofNullable<Error>(this.throwable); }

    isSuccess(): boolean { return this.output != null; }

    isFailure(): boolean { return this.throwable != null; }

    static of<I,O>(method: string,input: I,output: O): Mapped<I,O>;
    static of<I,O>(method: string,input: I,throwable: Error): Mapped<I,O>;
    static of<I,O>(method: string,input: I,third: O | Error): Mapped<I,O>
    {
        if (method == null)
            throw new NullPointerException("method is null");

        if (input == null)
            throw new NullPointerException("input is null");

        if (third instanceof Error)
            return new Mapped<I,O>(method,input,null,third);

        if (third == null)
            return new Mapped<I,O>(method,input,null,new NullPointerException("output is null"));

        return new Mapped<I,O>(method,input,third,null);
    }
}

//////////////////////////////////////////////////////////////////////////////

File: Components/Strata.Foundation.Core/main/java/strata/foundation/core/utility/Mapped.java
//////////////////////////////////////////////////////////////////////////////
// Mapped.java
//////////////////////////////////////////////////////////////////////////////

package strata.foundation.core.utility;

import java.util.Objects;
import java.util.Optional;

/**
 * <p>
 * Result produced by an operation that accepts an input and returns
 * an output, capturing the method name, the input value, and either
 * the resulting output or the {@link Throwable} thrown while
 * producing it. This covers any transform scenario —
 * {@link java.util.function.Function} applications, type
 * converters, serializers, parsers, or any method that maps one
 * value to another. Exactly one of output or exception is present;
 * {@link #isSuccess()} and {@link #isFailure()} indicate which
 * outcome occurred.
 * </p><br/>
 * <b>Examples</b><br/>
 * <pre>
 * Mapped&lt;String,Integer&gt; result =
 *     Mapped.of("parseInt","42",Integer.parseInt("42"));
 *
 * if (result.isSuccess())
 * {
 *     Integer value = result.getOutput().get();
 * }
 * </pre>
 *
 * @param <I> - the type of the input value passed to the operation
 * @param <O> - the type of the output value produced by the operation
 */
public
class Mapped<I,O>
{
    private final String    method;
    private final I         input;
    private       O         output;
    private       Throwable throwable;

    public
    Mapped(String method,I input,O output)
    {
        this.method = Objects.requireNonNull(method,"method");
        this.input  = Objects.requireNonNull(input,"input");

        try
        {
            this.output =
                Objects.requireNonNull(output,"output");
            this.throwable = null;
        }
        catch (Exception e)
        {
            this.output    = null;
            this.throwable = e;
        }
    }

    public
    Mapped(String method,I input,Throwable throwable)
    {
        this.method = Objects.requireNonNull(method,"method");
        this.input  = Objects.requireNonNull(input,"input");
        this.output = null;

        try
        {
            this.throwable =
                Objects.requireNonNull(throwable,"throwable");
        }
        catch (Exception e)
        {
            this.throwable = e;
        }
    }

    public String
    getMethod() { return method; }

    public I
    getInput() { return input; }

    public Optional<O>
    getOutput() { return Optional.ofNullable(output); }

    public Optional<Throwable>
    getException() { return Optional.ofNullable(throwable); }

    public boolean
    isSuccess() { return this.output != null; }

    public boolean
    isFailure() { return this.throwable != null; }

    public static <I,O> Mapped<I,O>
    of(String method,I input,O output) { return new Mapped<I,O>(method,input,output); }

    public static <I,O> Mapped<I,O>
    of(String method,I input,Throwable throwable) { return new Mapped<I,O>(method,input,throwable); }
}

//////////////////////////////////////////////////////////////////////////////
